### PR TITLE
Persist device id on volume

### PR DESCRIPTION
**Pull request by robiscoding**
# Description

- Persisting DEVICE_ID so that we track monitoring data by device id more reliably

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [X] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

- Set DEVICE_ID as env var; build and run container without a volume; print the device ID; re-run with different DEVICE_ID and observe that the DEVICE_ID changes
- Re-running with changes on this branch prevents the device ID from changing

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: inference/core/devices/utils.py [hunk 1]
```diff
@@ -136,5 +137,27 @@ def get_inference_server_id():
         return "UNKNOWN"
 
 
+def get_global_device_id():
+    cache_dir = f"{MODEL_CACHE_DIR}/_config"
```
Comment (grzegorz-roboflow) [inline inference/core/devices/utils.py:141]: `os.path.join` is used in other parts of the codebase, lets use it here too

File: inference/core/devices/utils.py [hunk 2]
```diff
@@ -136,5 +137,27 @@ def get_inference_server_id():
...
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir)
+    else:
+        with open(cache_dir, "r") as f:
+            config = json.loads(f.read())
+            if config.get("global_device_id") is not None:
+                return config["global_device_id"]
+    global_device_id = DEVICE_ID if DEVICE_ID is not None else get_device_hostname()
+    with open(cache_dir, "w") as f:
```
Comment (grzegorz-roboflow) [inline inference/core/devices/utils.py:150]: There is a low risk of ending up with garbage in that file due to race condition. We could consider locking https://docs.python.org/3/library/fcntl.html#fcntl.flock

```python
with open(cache_dir, "w") as f:
        fcntl.flock(f, fcntl.LOCK_EX)
        try:
            # (...)
        except Exception as exc:
            # unsuccessful write attempt
        fcntl.flock(f, fcntl.LOCK_UN)
```

File: inference/core/devices/utils.py [hunk 3]
```diff
@@ -136,5 +137,27 @@ def get_inference_server_id():
...
+        f.write(
+            json.dumps(
```
Comment (grzegorz-roboflow) [inline inference/core/devices/utils.py:152]: we can use `json.dump` instead (https://docs.python.org/3/library/json.html#json.dump)

**Review by grzegorz-roboflow — CHANGES_REQUESTED** [review]

**Comment by robiscoding** [comment]
Will re-open when I have bandwidth to actively work on this

**Pull request closed without merging**